### PR TITLE
[WIP] Issue{40,44} | SlideforMap changes

### PR DESCRIPTION
**Pull request by haydn-jones**
See description of #44

### CODE REVIEW

**Comment by deniscohen** [comment]
@haydn-jones Can you illustrate a call for that function? 

Can we capitalize `computeSlope` to `ComputeSlope` (I know I have my own idiosyncrasy).

**Comment by deniscohen** [comment]
@haydn-jones Should we include the optional unit for the output raster for slope: `deg` or `rad` or `m/m`?
So the call would have 3 args: raster, enum_value, something_for_units.

File: KiLib/Raster/Raster.cpp [hunk 1]
```diff
@@ -155,4 +156,23 @@ namespace KiLib
...
+      std::map<KiLib::Raster::SlopeMethod, std::function<KiLib::Raster(const KiLib::Raster &)>>{
+         {KiLib::Raster::SlopeMethod::ZevenbergenThorne, KiLib::Raster::computeSlopeZevenbergenThorne},
+      };
+
+   KiLib::Raster KiLib::Raster::computeSlope(KiLib::Raster::SlopeMethod method)
+   {
+      return EnumToSlope.at(method)(*this);
+   }
+
+   KiLib::Raster KiLib::Raster::computeSlopeZevenbergenThorne(const KiLib::Raster &inp)
```
Comment (oparkins) [inline KiLib/Raster/Raster.cpp:170]: Are you returning the Raster so that you can do some method chaining?

File: KiLib/Raster/Raster.hpp [hunk 2]
```diff
@@ -205,6 +205,15 @@ namespace KiLib
...
       }
 
+      // Slope methods
+      enum SlopeMethod
+      {
+         ZevenbergenThorne,
+      };
+
+      KiLib::Raster        computeSlope(KiLib::Raster::SlopeMethod method);
+      static KiLib::Raster computeSlopeZevenbergenThorne(const KiLib::Raster &inp);
```
Comment (oparkins) [inline KiLib/Raster/Raster.hpp:215]: This comment could very likely be from lack of knowledge and ignorance:
- Should a slope calculation function be returning a raster? Seems that it should be returning just a double or something. (uninformed dumb thought) 
- Or maybe the use case is that it doesn't modify the Raster object and you need to return the Raster? (I like functions that don't modify the initial obj tbh.) If so, maybe a `const` is in order?

**Review by oparkins — COMMENTED** [review]
Other than my lack of specific knowledge of the use case for this, it looks good. Nice and elegant for what it seems it is trying to accomplish. Mostly my comments are from my lack of understanding of the use case.

File: KiLib/Raster/Raster.hpp (hunk 2 above)
Comment (deniscohen) [reply KiLib/Raster/Raster.hpp:215]: @haydn-jones Following Owen's second command: Is the input raster modified? It should not. I am not sure whether it is.

**Comment by haydn-jones** [comment]
> @haydn-jones Can you illustrate a call for that function?
> 
> Can we capitalize `computeSlope` to `ComputeSlope` (I know I have my own idiosyncrasy).

Assuming we have a DEM raster called `rast` it would be: `KiLib::Raster slope = rast.computeSlope(ENUM_VALUE_HERE)`

File: KiLib/Raster/Raster.cpp (hunk 1 above)
Comment (haydn-jones) [reply KiLib/Raster/Raster.cpp:170]: I believe we want a new raster that holds the slope at every point in the provided raster.

File: KiLib/Raster/Raster.hpp (hunk 2 above)
Comment (haydn-jones) [reply KiLib/Raster/Raster.hpp:215]: @oparkins It should return a raster, its computing a value of slope at every point.

@deniscohen I'll add a const to `KiLib::Raster        computeSlope(KiLib::Raster::SlopeMethod method);`, the actual slope computations take a const ref so we're good there.

**Comment by oparkins** [comment]
Also, I'm going to change this to merge into `dev`

File: KiLib/Raster/Raster.cpp (hunk 1 above)
Comment (deniscohen) [reply KiLib/Raster/Raster.cpp:170]: @haydn-jones Correct. We want a raster

**Review by deniscohen — COMMENTED** [review]
Design looks good to me

**Review by deniscohen — APPROVED** [review]

**Comment by deniscohen** [comment]
@haydn-jones @oparkins I don't know why Windows is not building.

**Comment by deniscohen** [comment]
> @haydn-jones Can you illustrate a call for that function?
> 
> Can we capitalize `computeSlope` to `ComputeSlope` (I know I have my own idiosyncrasy).

Done in 50370d89359145805d5dc3e56804bc7d483655c4

**Comment by deniscohen** [comment]
@haydn-jones @oparkins Except for Windows not building, I think this is ready to merge. If the Windows problem cannot be solved quickly let's still merge as we need this to move forward with PRIMULA and SlideforMAP. Thanks

**Pull request merged**